File: src/location/LocationRecord.java
package location;

import location.databinding.PlainRecordObject;

/**
 * A comparable, and therefore sortable location record class
 * Sorting is based purely on the timestamps of the records
 */
public class LocationRecord implements Comparable<LocationRecord> {

    private long timeStampMS;
    private double latitude;
    private double longitude;
    private int accuracy;

    /**
     * Ctor only for temporary records, does not store actual data!
     *
     * @param timeMS the timestamp to store, in milliseconds
     */
    LocationRecord(long timeMS) {
        this.timeStampMS = timeMS;
        this.latitude = 0;
        this.longitude = 0;
        this.accuracy = 0;
    }

    /**
     * Ctor from a POJO location record
     *
     * @param po the record to get the data from
     */
    LocationRecord(PlainRecordObject po) {
        this.timeStampMS = po.timestampMs;
        this.latitude = convertIntLatLong(po.latitudeE7);
        this.longitude = convertIntLatLong(po.longitudeE7);
        this.accuracy = po.accuracy;
    }

    /**
     * Helper function to turn integer formatted latlong coordinates into doubles
     *
     * @param latlong the coordinate to convert
     * @return the coordinate as a double value
     */
    private double convertIntLatLong(int latlong) {
        return (double) latlong * 0.0000001;
    }

    public long getTimeStampMS() {
        return timeStampMS;
    }

    public double getLatitude() {
        return latitude;
    }

    public double getLongitude() {
        return longitude;
    }

    public int getAccuracy() {
        return accuracy;
    }

    @Override
    public int compareTo(LocationRecord o) {
        return Long.compare(timeStampMS, o.timeStampMS);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj.getClass() != LocationRecord.class) {
            return false;
        }

        LocationRecord rec = (LocationRecord) obj;
        return timeStampMS == rec.timeStampMS;
    }
}

File: src/main/kotlin/location/LocationRecord.kt
package location

import location.databinding.PlainRecordObject

class LocationRecord @JvmOverloads constructor(
        val timeStampMS: Long,
        val latitude: Double = 0.0,
        val longitude: Double = 0.0,
        val accuracy: Int = 0)
    : Comparable<LocationRecord> {

    companion object {

        /**
         * Converts integer latlong coordinates into doubles
         * e.g. 135590300 to 13.5590300
         */
        private fun Int.asCoordinate() = this * 0.0000001

        @JvmStatic
        fun from(plainRecordObject: PlainRecordObject) =
                LocationRecord(
                        plainRecordObject.timestampMs,
                        plainRecordObject.latitudeE7.asCoordinate(),
                        plainRecordObject.longitudeE7.asCoordinate(),
                        plainRecordObject.accuracy
                )
    }

    override fun compareTo(other: LocationRecord) = timeStampMS.compareTo(other.timeStampMS)

}
